File: src/my-components/cell/textCell.ts
import {
  CellPlugin,
  BasePayload,
} from "./cellPluginSystem"


export const TextCellPlugin: CellPlugin = {
  type: "Text",

  getDefaultPayload: (): BasePayload => ({
    value: "",
    label: "",
  }),

  render: (ctx, cell, x, y, _w, h, options) => {
    if (cell.type === "Text") {

      let str = String(cell.payload.value ?? "");
      if (str.length > 6) str = str.slice(0, 6) + "..";

      let paddingX = options?.paddingX ?? 4;

      // 🔥 關鍵：文字樣式設定一定要放 plugin 裡
      ctx.fillStyle = options?.textColor ?? "#000";
      ctx.font = options?.font ?? "14px sans-serif";
      ctx.textAlign = "left";
      ctx.textBaseline = "middle";

      ctx.fillText(str, x + paddingX, y + h / 2);

    } 
    else
      throw new Error(`this cell type:${cell.type} is not Text!`);

  }

}

File: src-tauri/src/cell_plugins/text_cell.rs
use std::collections::HashMap;

use crate::cell_plugins::{cell::BasePayload};

use super::CellPlugin;
use serde::{Deserialize, Serialize};
use serde_json::{json};

pub struct TextCellPlugin;

#[derive(Debug, Deserialize, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct TextCellPayload {
    pub value: String,
    pub label: Option<String>
}



impl CellPlugin for TextCellPlugin {
    fn get_type_id(&self) -> &str { "Text" }

    fn get_display_name(&self) -> &str { "文字-Text" }

    fn display_cell(&self, payload: BasePayload) -> String {
        let val = payload.value;
        return val.to_string();       
    }    

    fn default_payload(&self) -> Result<BasePayload, String> {

        Ok(BasePayload {
            value: json!("".to_string()),
            display_style: None,
            display_value: None,
            extra_fields: HashMap::new(),
        })
    }

    fn get_css(&self) -> String {
        ".cell-plugin-text {
            font-size: 14px;
            font-family: 'Segoe UI', 'Roboto', sans-serif;
            color: #222;
            white-space: nowrap;
            overflow: hidden;
            text-overflow: ellipsis;
            line-height: 1.4;
        }"
        .to_string()
    }

}